zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -0,0 +1,521 @@
+            OpcodeId::PC => ExecutionState::PC,
+            _ => unimplemented!("unimplemented opcode {:?}", step.op),
+        }
+    }
+}
+
+impl From<&bus_mapping::bytecode::Bytecode> for Bytecode {
+    fn from(b: &bus_mapping::bytecode::Bytecode) -> Self {
+        Bytecode::new(b.to_bytes())
+    }
+}
+
+fn step_convert(
+    prev: Option<&bus_mapping::circuit_input_builder::ExecStep>,
+    step: &bus_mapping::circuit_input_builder::ExecStep,
+    ops_len: (usize, usize, usize),
+) -> ExecStep {
+    let (stack_ops_len, memory_ops_len, _storage_ops_len) = ops_len;
+    let result = ExecStep {
+        rw_indices: step
+            .bus_mapping_instance
+            .iter()
+            .map(|x| {
+                let index = x.as_usize() - 1;
+                match x.target() {
+                    bus_mapping::operation::Target::Stack => index,
+                    bus_mapping::operation::Target::Memory => {
+                        index + stack_ops_len
+                    }
+                    bus_mapping::operation::Target::Storage => {
+                        index + stack_ops_len + memory_ops_len
+                    }
+                }
+            })
+            .collect(),
+        execution_state: ExecutionState::from(step),
+        rw_counter: usize::from(step.gc),
+        program_counter: usize::from(step.pc) as u64,
+        stack_pointer: 1024 - step.stack_size,
+        gas_left: step.gas_left.0,
+        gas_cost: step.gas_cost.as_u64(),
+        opcode: Some(step.op),
+        // As in https://github.com/ethereum/go-ethereum/blob/bc6bf1e1937829b5d5b0d431a9333d47c3e08082/core/vm/interpreter.go#L224-L233,
+        // geth increases memory size before making trace,
+        // so the memory size in ExecStep is not what we expect to see.
+        // We have to use the memory size of previous step as correct memory
+        // size before each step
+        memory_size: match prev {
+            None => 0,
+            Some(prev_step) => (prev_step.memory_size as u64) / 32, /* memory size in word */
+        },
+        ..Default::default()
+    };
+    result
+}
+
+fn tx_convert(
+    randomness: Fp,
+    bytecode: &Bytecode,
+    tx: &bus_mapping::circuit_input_builder::Transaction,
+    ops_len: (usize, usize, usize),
+) -> Transaction<Fp> {
+    let mut result: Transaction<Fp> = Transaction::<Fp> {
+        calls: vec![Call {
+            id: 1,
+            is_root: true,
+            is_create: tx.is_create(),
+            opcode_source: RandomLinearCombination::random_linear_combine(
+                bytecode.hash.to_le_bytes(),
+                randomness,
+            ),
+        }],
+        ..Default::default()
+    };
+    result.calls = vec![Call {
+        id: 1,
+        is_root: true,
+        is_create: tx.is_create(),
+        opcode_source: RandomLinearCombination::random_linear_combine(
+            bytecode.hash.to_le_bytes(),
+            randomness,
+        ),
+    }];
```

Comment:
Isn't this a repetition of the previous declaration of `calls` in the declaration of `result`?

Reply:
removed in https://github.com/appliedzkp/zkevm-circuits/pull/239/commits/346da63f8292f79bb9687bd3471bf3ebc783435e